src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,652 @@
+    /// @return requestId The index of the withdrawal request
+    /// @return assets The amount of liquidity assets that will be claimable by the redeemer
+    function requestRedeem(uint256 shares) external returns (uint256 requestId, uint256 assets) {
+        // Calculate the amount of assets to transfer to the redeemer
+        assets = convertToAssets(shares);
+
+        requestId = nextWithdrawalIndex;
+        uint120 queued = SafeCast.toUint120(withdrawsQueued + assets);
+        uint40 claimTimestamp = uint40(block.timestamp + CLAIM_DELAY);
+
+        // Store the next withdrawal request
+        nextWithdrawalIndex = SafeCast.toUint16(requestId + 1);
+        // Store the updated queued amount which reserves WETH in the withdrawal queue
+        withdrawsQueued = queued;
+        // Store requests
+        withdrawalRequests[requestId] = WithdrawalRequest({
+            withdrawer: msg.sender,
+            claimed: false,
+            claimTimestamp: claimTimestamp,
+            assets: SafeCast.toUint120(assets),
+            queued: queued
+        });
+
+        // burn redeemer's shares
+        _burn(msg.sender, shares);
+
+        // Remove the redeemed assets from the last available assets
+        lastAvailableAssets -= SafeCast.toInt128(SafeCast.toInt256(assets));
+
+        emit RedeemRequested(msg.sender, requestId, assets, queued, claimTimestamp);
+    }
+
+    /// @notice Claim liquidity assets from a previous withdrawal request after the claim delay has passed
+    /// @param requestId The index of the withdrawal request
+    /// @return assets The amount of liquidity assets that were transferred to the redeemer
+    function claimRedeem(uint256 requestId) external returns (uint256 assets) {
+        // Load the structs from storage into memory
+        WithdrawalRequest memory request = withdrawalRequests[requestId];
+
+        require(request.claimTimestamp <= block.timestamp, "Claim delay not met");
+        // Is there enough liquidity to claim this request?
+        require(
+            request.queued <= withdrawsClaimed + IERC20(liquidityAsset).balanceOf(address(this)),
+            "Queue pending liquidity"
+        );
+        require(request.withdrawer == msg.sender, "Not requester");
+        require(request.claimed == false, "Already claimed");
+
+        // Store the request as claimed
+        withdrawalRequests[requestId].claimed = true;
+        // Store the updated claimed amount
+        withdrawsClaimed += request.assets;
+
+        assets = request.assets;
+
+        emit RedeemClaimed(msg.sender, requestId, assets);
+
+        // transfer the liquidity asset to the withdrawer
+        IERC20(liquidityAsset).transfer(msg.sender, assets);
+    }
+
+    /// @notice Check if a withdrawal request can be claimed
+    /// @param requestId The index of the withdrawal request
+    function isClaimable(uint256 requestId) public view returns (bool) {
```

Comment:
🟠 After chewing it around more, I do prefer a view method for the claimable amount, instead of an isClaimable method. Two reasons:

- We are duplicating logic with the current logic. :(
- With `isClaimable`, the UI has to make one request per withdraw request. With the claimable counter that's only one request and then the UI can know if unlimited numbers of claims are ready to go. Also makes it so we can reuse the old UI code.

Reply:
Good point. I've added a `claimable` view function 

```Solidity
/// @notice Used to work out if an ARM's withdrawal request can be claimed.
/// If the withdrawal request's `queued` amount is less than the returned `claimable` amount, then if can be claimed.
function claimable() public view returns (uint256) {
    return withdrawsClaimed + IERC20(liquidityAsset).balanceOf(address(this));
}
```